pkg/scheduler/scheduler_test.go

```diff
@@ -82,6 +143,93 @@ func TestCreateScheduler(t *testing.T) {
 			if tc.enableSchedulerEstimator && tc.schedulerEstimatorPort != sche.schedulerEstimatorClientConfig.TargetPort {
 				t.Errorf("unexpected schedulerEstimatorPort want %v, got %v", tc.schedulerEstimatorPort, sche.schedulerEstimatorClientConfig.TargetPort)
 			}
+
+			if tc.disableSchedulerEstimatorInPullMode != sche.disableSchedulerEstimatorInPullMode {
+				t.Errorf("unexpected disableSchedulerEstimatorInPullMode want %v, got %v", tc.disableSchedulerEstimatorInPullMode, sche.disableSchedulerEstimatorInPullMode)
+			}
+
+			if tc.schedulerEstimatorServicePrefix != sche.schedulerEstimatorServicePrefix {
+				t.Errorf("unexpected schedulerEstimatorServicePrefix want %v, got %v", tc.schedulerEstimatorServicePrefix, sche.schedulerEstimatorServicePrefix)
+			}
+
+			if tc.schedulerName != sche.schedulerName {
+				t.Errorf("unexpected schedulerName want %v, got %v", tc.schedulerName, sche.schedulerName)
+			}
+
+			if tc.enableEmptyWorkloadPropagation != sche.enableEmptyWorkloadPropagation {
+				t.Errorf("unexpected enableEmptyWorkloadPropagation want %v, got %v", tc.enableEmptyWorkloadPropagation, sche.enableEmptyWorkloadPropagation)
+			}
+		})
+	}
+}
+
+func TestReconcileEstimatorConnection(t *testing.T) {
+	dynamicClient := dynamicfake.NewSimpleDynamicClient(runtime.NewScheme())
+	karmadaClient := karmadafake.NewSimpleClientset()
+	kubeClient := fake.NewSimpleClientset()
+	scheduler, err := NewScheduler(dynamicClient, karmadaClient, kubeClient, WithEnableSchedulerEstimator(true))
+	if err != nil {
+		t.Fatalf("Failed to create scheduler: %v", err)
+	}
+
+	clusterName := "test-cluster"
+	cluster := &clusterv1alpha1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{Name: clusterName},
+		Spec:       clusterv1alpha1.ClusterSpec{SyncMode: clusterv1alpha1.Pull},
+	}
+	_, err = karmadaClient.ClusterV1alpha1().Clusters().Create(context.TODO(), cluster, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to create cluster: %v", err)
+	}
+
+	tests := []struct {
+		name                                string
+		key                                 util.QueueKey
+		disableSchedulerEstimatorInPullMode bool
+		expectedError                       bool
+		expectEstablishConnection           bool
+	}{
```

Comment:
It seems that all cases have an error when getting the cluster object.

Reply:
Hello @XiShanYongYe-Chang I think the issue is that test is creating a cluster using karmadaClient, but the reconcileEstimatorConnection method is using a clusterLister to retrieve it. This can be a reason of an error when getting the cluster object.  More cases like the cluster in push mode might help us determine it. There can be also other reasons of this errors. @varshith257

Reply:
I have tried though and debuged it. Maybe I can add this test in separate PR with fresh perspective of solving it. I will remove adding this test in this PR.

Reply:
> I think the issue is that test is creating a cluster using karmadaClient, but the reconcileEstimatorConnection method is using a clusterLister to retrieve it.

@anujagrawal699 I see you wrote a test with similar logic and it passed, so wanted to see what you thought.

> I have tried though and debuged it. Maybe I can add this test in separate PR with fresh perspective of solving it. I will remove adding this test in this PR.

Hi @varshith257 I agree with you.

Reply:
@XiShanYongYe-Chang In desheduler_test.go, the tests starts and waits for the informer cache to sync and also it allows the test operations to complete with timeout. Also, i don't think the test is successfully retrieving the cluster.` _, err = karmadaClient.ClusterV1alpha1().Clusters().Create(context.TODO(), cluster, metav1.CreateOptions{})` is creating a cluster. I think retrieving it using the same clusterLister method could work, it goes like```
 retrievedCluster, err := scheduler.clusterLister.Get(clusterName)
if err != nil {
    t.Fatalf("Failed to retrieve cluster: %v", err)
}
```